### PR TITLE
Turn the game class into a proper singleton and fix associated crash

### LINKED ISSUE #16745: "Simplified nutrition" mod crashes the game when entering character creation
When I create a world that has the simplified nutrition mod activated, attempting to create a character for that world immediately crashes the game, with a generic "cataclysm-tiles.exe has stopped working" message. Version: 0.C-14177-gae51d65

### LINKED ISSUE #16827: Crash when Creating world with Perfect Reliability mod installed
#On a fresh install, when ever I enabled the Perfect Reliability mod the game crashed when  it generated the world.

### PR DESCRIPTION
**Pull request by codemime**
Fixes #16745, fixes #16827. 

`game::get_instance()` can now be called safely from almost any line of the code and returns a reference (no messing with pointers in the future). To minimize the impact I left global `g` variable unchanged except the line where it caused the said crash. Also, there was a memory leak in the tests (`g` was never `delete`d).

I believe that resetting (for further reuse) should be separated from destroying in the future, so that singletons could invoke each other during resetting, but destroy themselves separately and independently.

### CODE REVIEW

File: src/game.h [hunk 1]
```diff
@@ -20,7 +20,7 @@ extern int save_loading_version;
 
 // The reference to the one and only game instance.
 class game;
-extern game *g;
+extern game *g; // @todo replace all occurrences with 'game::get_instance()' and delete the global variable
```
Comment (Coolthulhu) [inline src/game.h]: It will need some shorter name than `game::get_instance()`.

**Comment by BevapDin** [comment]
I am for moving stuff out of the game class (instead of this change).

For example the map could be a global static, as could be the player instance, the UI stuff (WINDOW*, live_view) could go into a separate GUI class. The creature tracker could be exposed directly etc.

File: src/game.h (hunk 1 above)
Comment (codemime) [inline src/game.h]: What would you prefer?

**Comment by codemime** [comment]
> I am for moving stuff out of the game class (instead of this change)

Me too, but **along** with this change (we are going to keep the `game` class, aren't we?). I don't see why instead? How they are opposite?

> For example the map could be a global static

Agreed, but yet another singleton.

> as could be the player instance

Quite arguable, since (as far as I know) we plan to introduce some kind of multilayer (probably in a distant future). It just can't be a single instance.

> the UI stuff (WINDOW*, live_view) could go into a separate GUI class. 

Definitely yes! But how does it connected to this change? I still don't get it.

File: src/game.h (hunk 1 above)
Comment (Coolthulhu) [inline src/game.h]: A macro that invokes it, a global function or just `game::get()`.

**Comment by BevapDin** [comment]
Why not declare `g` like this:
```C++
extern game &g; // in the header.

static game &game_instance() {
    static game instance;
    return instance;
}
game &g = game_instance();
```
You may have to adjust some of the constructors involved here, but this would keep the global `g` available as it is.

>> as could be the player instance

> Quite arguable, since (as far as I know) we plan to introduce some kind of multiplayer (probably in a distant future). It just can't be a single instance.

Well, the player instance is already a global static. Only it's not name `g->u` instead of just `u`. As long as there is only one game instance `g`, there is only one player instance `u`, so what's the point of putting it into `g`?

**Comment by codemime** [comment]
> Why not declare g like this:

It's hardly any better than our current implementation (except that this doesn't use `new` and `delete`). The whole point of the singleton is that it gets created on demand when stars are right. You simply don't need this line: `game &g = game_instance();` because it's a headache where to put it (is `g` initialized already or not?). Just one lapse and we get non-obvious undefined behaviour (just like the crash that this PR fixes, but we're lucky that it crashes). Singletons are awful, but they're far better than global variables.

> As long as there is only one game instance g, there is only one player instance u

Technically there are as many instances of `player`, as there are actual players. `u` is a placeholder in that case (or an heir, or a wrapper may be. doesn't matter). It has nothing to do with a single instance.

> so what's the point of putting it into g?

No point at all, but I never told anything about that. Again, how's moving stuff out of the `game` class connected with this change?

File: src/init.cpp [hunk 2]
```diff
@@ -348,7 +349,6 @@ void DynamicDataLoader::unload_data()
     SNIPPET.clear_snippets();
     vehicle_prototype::reset();
     vpart_info::reset();
-    MonsterGenerator::generator().reset();
```
Comment (BevapDin) [inline src/init.cpp]: Why was this removed? Are the monster types and species cleared somewhere else?

**Comment by BevapDin** [comment]
> You simply don't need this line:  game &g = game_instance();  because it's a headache where to put it (is  g  initialized already or not?). 

`g` is a static variable, it's initialized before `main` is run. The compiler issues the initialization exactly once (it's a reference). The function that returns the reference creates a game object statically, the compiler (again) will ensure it's created only once.

Granted, this involves inspecting / checking that the involved constructors don't actually do much and just set up the object itself and don't access anything outside of them.

But you're right. Considering the current state of the code (`g` being a pointer and being used everywhere), your fix is much more reasonable.

> Technically there are as many instances of  player , as there are actual players.  u  is a placeholder in that case (or an heir, or a wrapper may be. doesn't matter). It has nothing to do with a single instance.

Are we talking about what should be or what is? Because currently there is only one `g->u` because there is only one `g`. A few function have a dedicated player parameter, but many still use `g->u`, especially when it's about input (from keyboard) or for printing messages.

> Again, how's moving stuff out of the  game  class connected with this change?

It fixes the same problem, but with in different and more laborious way (-;
Your fix is fine as well.

**Comment by mugling** [comment]
The two approaches are not orthogonal. `game` is about the only class that is unarguably a singleton however splitting some of our overweight classes up would also be good.

**Comment by kevingranade** [comment]
The problem here is that options are being stashed in the game object, not that the game object isn't immortal.  We already have per-world options and per-player option mechanisms, why isn't it using one of those?

Singletons are one of the only programming constructs worse than globals, this is a regression for code quality, not an improvement.

File: src/init.cpp (hunk 2 above)
Comment (codemime) [inline src/init.cpp]: Yes, [here](https://github.com/CleverRaven/Cataclysm-DDA/pull/17093/files#diff-18513665750ef5adf42b5ec29e14162eR323). This line was moved because of the problem I mentioned both in the description and the comment: singletons should be destroyed independently, so that order doesn't matter. With this line here, the `game` class tries to reset the already destroyed object.

**Comment by codemime** [comment]
> g is a static variable, it's initialized before main is run. The compiler issues the initialization exactly once (it's a reference). The function that returns the reference creates a game object statically, the compiler (again) will ensure it's created only once.

Oh, sorry, I misread your proposal. You were talking about static linking. That makes sense, but I have certain doubts that it will work fine with our current code without reworking. Also, it has its drawbacks too.

> Are we talking about what should be or what is?

What should be. I was talking about multipleer (see above). 

> It fixes the same problem, but with in different and more laborious way (-;

I totally agree with you about moving stuff out of the `game` class. @mugling confirmed what I've been trying to express: these things are not opposed.

**Comment by codemime** [comment]
> The problem here is that options are being stashed in the game object, not that the game object isn't immortal. We already have per-world options and per-player option mechanisms, why isn't it using one of those?

Good point. @mugling could answer the question better than me. I suppose he had his reasons.

> Singletons are one of the only programming constructs worse than globals (of which they are a subset), this is a regression for code quality, not an improvement.

Well, it's already a singleton whether we agree with that or not. I've just actualized the existing thing:

1. Accessed globally
2. Only one instance is possible (try to create and destroy yet another one, and it will crash)
3. Lives long and dies on program termination, not earlier.

Global access (instead of passing things around) is a bad thing, there's no doubt. But how are singletons worse at this point than global variables, especially than fragile global pointers?

File: src/game.cpp [hunk 3]
```diff
@@ -344,7 +352,6 @@ void game::load_data_from_dir(const std::string &path)
 
 game::~game()
 {
-    DynamicDataLoader::get_instance().unload_data();
```
Comment (BevapDin) [inline src/game.cpp]: I suspect all those unload / reset functions don't need to be called in the destructor at all. Things should clear themself and nothing outside.

`mapbuffer::~mapbuffer` already class its reset method, so it's not needed here.
`DynamicDataLoader::unload_data` shouldn't be needed at all (in any destructor), the game data should be cleared via the destructor of the various containers. It is needed only before loading the game data or after a game has been saved / lost (not related to the program end at all).

`world_generator` and `gamemode` should be `unique_ptr` and thereby deleting themself as needed.

File: src/game.cpp [hunk 4]
```diff
@@ -319,11 +320,18 @@ void game::load_core_data()
     // core data can be loaded only once and must be first
     // anyway.
     DynamicDataLoader::get_instance().unload_data();
+    MonsterGenerator::generator().reset(); // @todo Separate clearing from destroying. It's not the best place for keeping it.
```
Comment (BevapDin) [inline src/game.cpp]: Regarding the comment `reset` here resets the content, it does not destruct the generator. This is already separate: destruction happens when the static instance in `MonsterGenerator::generator` is destructed at the end of the program.

`reset` should be called from `unload_data` there is no reason to move this out of there. But `unload_data` should not be called from `game::~game`.

File: src/game.cpp (hunk 3 above)
Comment (codemime) [inline src/game.cpp]: That's exactly why I removed this line from the destructor.

> Things should clear themself and nothing outside.

That's what I meant by:

> but destroy themselves separately and independently.

File: src/game.cpp (hunk 4 above)
Comment (codemime) [inline src/game.cpp]: > Regarding the comment reset here resets the content, it does not destruct the generator.

And that's a good thing, because it's in `game::load_core_data()`, not in a destructor.

> reset should be called from unload_data there is no reason to move this out of there.

There [is](https://github.com/CleverRaven/Cataclysm-DDA/pull/17093/files#r66237782) a reason:

> With this line here, the game class tries to reset the already destroyed object.

**Comment by kevingranade** [comment]
On Jun 8, 2016 4:36 AM, "codemime" <notifications@github.com> wrote:
>> Singletons are one of the only programming constructs worse than globals
(of which they are a subset), this is a regression for code quality, not an
improvement.
>
> Well, it's already a singleton whether we agree with that or not. I've
just actualized the existing thing:
>
> Accessed globally
> Only one instance is possible (try to create and destroy yet another one,
and it will crash)
> Lives long and dies on program termination, not earlier.

Precisely, the only thing you've done is the second plus extending the
third a bit. Extending the lifespan earlier is the actual fix here, but
object lifespans starting at some indeterminate time early in program
startup is the problem, not something we need more of.

> Global access (instead of passing things
> around) is a bad thing, there's no doubt. But
> how are singletons worse at this point than
> global variables, especially than fragile global
> pointers?

We have no need to *enforce* the existence of only one game object, in fact
it prevents us from doing the actual right thing, which is nesting
everything under game and destroying/recreating it if the player exits to
the main menu and starts/loads a new game. This would allow us to do all of
the "resetting" in destructors instead of trying to sequence them manually.

Ideally we'd have a single explicitly constructed object hierarchy rooted
in main.  One of the problems with singletons is "more singletons" quickly
becomes the most obvious solution to the problems created by your existing
singletons, this issue is a perfect example.
